Location saving

SimplyLinn opened the pull request:
Resolves #108
Uhm, yeah... I'm just gonna leave this here.

I've not written any tests, nor done any extensive testing myself.

But I'm not about to do so at the moment, this was just supposed to be "A quick thing" so I could get the functionality I wanted. It escalated into a complete mess of new and changed code and I hope it makes sense to you. I'm actually rather tired of this code at the moment 😛

Little did I know I had to start brushing up on my java-skills I haven't touched in years, using things I've never even seen before in Java. Fun times.

Anyway, sorry if my code is not up to standard, I'll understand if you want to discard my implementation and redo it yourself. I just thought, if you ended up in the same "WTF is this even?"-place as me, you'd might like to use the madness of another.

Anyway, this adds some more settings and preferences in the world.yml files. Namely, it dictates how and when to enforce a new location of a player when he/she teleports between worlds.

It also added a new MetaDataManager and a bunch of things I needed to attach information to a player object directly, as I didn't want to manually figure out when to discard data about players when they leave.
I tried to make it as generic as possible since I suspect you may want to do similar things in the future. If you'd like some documentation of how to use this mess, just ask! I'll be glad to explain.

(Although, you may want to do it before I forget my own code...)

Commit:
Added listener for player teleportation

Commit:
Port to using teleport event for world-change detection complete.

Commit:
Added weird .pom file some random atom-plugin generated in my project to gitignore

Commit:
Fixed up config files ond stuff

Commit:
Added settings to actual code.

Commit:
Started work on metadata implementation for hanging their last location in worlds onto the player object.

Commit:
Made a MetaDataFactory. Java is annoying sometimes. Just saying.

Commit:
Removed debug output. Sorry about last commit message, I was rather frustrated after trying to get something to work for a long time

Commit:
added the teleport-process. It works. Finally

Commit:
Removed debug message added in frustration, fixed the default worlds.yml to actually be usable by the code

MrOpposite commented:
I may consider trying to add tests... That coveralls report was rather depressing. But not now at 1AM. But don't count on it, I'm quite lazy when it comes to side projects 😛

ljacqu commented on src/main/java/me/gnat008/perworldinventory/PerWorldInventory.java:
```diff
@@ -156,6 +159,7 @@ protected void injectServices(Injector injector) {
         playerManager = injector.getSingleton(PWIPlayerManager.class);
         permissionManager = injector.getSingleton(PermissionManager.class);
         api = injector.getSingleton(PerWorldInventoryAPI.class);
+        metaDataManager = injector.getSingleton(PWIMetaDataManager.class);
```
You can undo this — the `metaDataManager` field is never used in this class.

ljacqu committed:
Fix formatting
- Add or remove various spaces
- Make fields private
- Use static methods via class, not via an instance
- Remove unused imports

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/metadata/PWIMetaDataManager.java:
```diff
@@ -0,0 +1,37 @@
+        return instance;
+    }
+
+    public <T> T getFromPlayer(Player player, String key) {
```
Similar to the comment above, it might makes sense here to have two (or three) individual methods for retrieving the values in a typed manner. This way you don't need to pass the `key` from outside—only this class needs to know about the key. 

Compare the current usage:
```java
// I need to pass "lastLocationInWorld" from every class that wants to use it
Map<String, Location> lastLocData = metaDataManager.getFromPlayer(player,"lastLocationInWorld");

// Also, the method is untyped, so nothing stops me from doing this:
Integer lastLocData = metaDataManager.getFromPlayer(player,"lastLocationInWorld"); // compiles!
```

vs. if a method `Map<String, Location> getLastLocationData(Player player)` existed:
```java
Map<String, Location> lastLocData = metaDataManager.getLastLocationData(player);
```

MrOpposite replied:
Yup, if I'm gonna make the manager explicitly aware of what implementations exists, that's the way to go. I was thinking in generic magic that only required adding a new implementation in a single file, and then it could be accessible trough the generic functions. In the beginning I had a bunch of more magic planned, but it broke down due to the inability to access static members of a class from a generic context. Since that didn't work, this is probably the way to go.

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/metadata/PWIMetaDataManager.java:
```diff
@@ -0,0 +1,37 @@
+    @Inject Injector injector;
+
+    PWIMetaDataManager() {}
+    public <T extends PWIMetaDataValueAbstract> T createMetadataValue(Class<T> clazz, Object... args) {
```
This is a lot of magic that can be simplified. You only have three (of which one is unused) implementations of `PWIMetaDataValueAbstract`. I'd suggest to have individual methods for creating these three classes, and to call the constructor manually.

```java
public PWIPlayerLastWorldInGroupData createLastWorldInGroupDataValue(Player player) {
   return new PWIPlayerLastWorldInGroupData(plugin, dataWriter, player);
}
```

i.e., 
1. Add `Player` to constructor of all PWIMetaDataValue classes where necessary
1. Remove `@Inject` from these constructors
1. Remove `init()` method; 
1. create individual methods in PWIMetaDataManager for creating metadata value classes

MrOpposite replied:
Yeah, I guess so. Will fix it up.

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/DataWriter.java:
```diff
@@ -67,4 +68,50 @@
+     * @param playerUUID The player unique id to save the last location in world for
+     * @param map The data to save
+     */
+    void saveLastLocationInWorld(UUID playerUUID, Map<String, Location> map);
```
Minor: `Map<String, Location> map` ->`Map<String, Location> locations`  ?

ljacqu commented on src/main/java/me/gnat008/perworldinventory/data/DataWriter.java:
```diff
@@ -67,4 +68,50 @@
+     * @param playerUUID The player unique id to save the last world in group for
+     * @param map The data to save
+     */
+    void saveLastWorldInGroup(UUID playerUUID, Map<String, String> map);
```
Minor: `Map<String, Location> map` ->`Map<String, Location> lastWorlds`  ?

ljacqu commented on src/main/java/me/gnat008/perworldinventory/groups/Group.java:
```diff
@@ -56,9 +60,41 @@ public Group(String name, List<String> worlds, GameMode gameMode) {
+     * @param usePosInWorld A {@link List} of teleport causes that should map to the last position in world
+     * @param configured If the group is defined in the worlds.yml file.
+     */
+    public Group(String name, List<String> worlds, List<String> useLastWorld, List<String> useLastPosInGroup, List<String> useLastPosInWorld, String defaultWorld, GameMode gameMode, boolean configured) {
```
The new `List<String>` fields, absolutely, forcibly, _must_ be enums. Nothing in the code suggests otherwise that this is only a set of predefined entries, and I've actually only started to understand this after having looked at the worlds.yml. 
All those entries should be of type `PlayerTeleportEvent.TeleportCause`.

MrOpposite replied:
Yup, very true. I did it this way to make it more obvious, but I guess it's, in retrospect, very silly.

I thought it'd make more sense if the set matched the external YML files datatype rather than the internal `TeleportCause`, but yeah... That's bonkers.

ljacqu reviewed (changes requested):
Nice! I've left some comments throughout the changes and submitted a PR to your fork to fix some of the formatting and member visibility.

One thing I haven't mentioned anywhere, but I see there's a lot of `PwiLogger.debug(String.format(...))` — String#format is quite slow, and even if debug is disabled this will be executed.

Commit:
Merge pull request #1 from ljacqu/LocationSaving

Small formatting changes

MrOpposite commented:
Ah yes, I was half way trough the process of rewriting the debugger in another branch, using String.format internally in the debug function. I decided against it after I realized my scope should be set to the location saving if I ever wanted to be finished, but some of the old String.format in debug is left from when I thought I'd make that the way to run it, so I'd just need to remove the string.format on my new debug-calls and all would be dandy.

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/groups/GroupManager.java:
```diff
@@ -162,6 +188,10 @@ public void saveGroupsToDisk() {
+            groupsConfigFile.set(groupKey + ".enforceLastWorld", group.shouldUseLastWorld());
+            groupsConfigFile.set(groupKey + ".enforceLastPosInGroup", group.shouldUseLastPosInGroup());
+            groupsConfigFile.set(groupKey + ".enforceLastPosInWorld", group.shouldUseLastPosInWorld());
+            groupsConfigFile.set(groupKey + ".enforceLastPosInWorld", group.getDefaultWorld());
```
Oh, totally... This was a last minute change when I noticed my naming convention in the YML file did not match.

Commit:
Fixed severe last-minute change of camelCase->kebab-case in the worlds.yml config

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/data/metadata/PWIMetaDataValueBasic.java:
```diff
@@ -0,0 +1,33 @@
+
+import javax.inject.Inject;
+
+public class PWIMetaDataValueBasic<T> extends PWIMetaDataValueAbstract<T> {
```
I left it there as a way to easily add primitive/simple data to players, for future uses. (In the beginning I was gonna use it for my data, but then I realized I actually could invalidate my data, and went with implementing it in another way)

Should this be removed entirely, or a way to add primitive data to players, using the `PWIMetaDataManager`, which I'm about to rewrite`?

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/data/metadata/PWIMetaDataManager.java:
```diff
@@ -0,0 +1,37 @@
+        MetadataValue data = null;
+        for(MetadataValue searchData : dataList) {
+            if(searchData.getOwningPlugin().equals(plugin)) {
+                data = searchData;
```
Clever, will do!

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/groups/GroupManager.java:
```diff
@@ -66,10 +86,10 @@ public void addGroup(String name, List<String> worlds) {
      * @param gamemode The default GameMode for this group.
      */
-    public void addGroup(String name, List<String> worlds, GameMode gamemode) {
+    public void addGroup(String name, List<String> worlds, List<String> useLastWorld, List<String> useLastPosInGroup, List<String> useLastPosInWorld, String defaultWorld, GameMode gamemode) {
```
Ah, thanks for the naming scheme. I've been staring at this code for a few days now and hit roadblock after roadblock in my coding, and since I need this functionality for my own server, the frustration was building up, making my "how to come up with good names"-part of my brain non-functional 😛

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/process/TeleportProcess.java:
```diff
@@ -0,0 +1,295 @@
+              player.getName(),
+              from.getWorld().getName(),
+              to.getWorld().getName()));
+            return;
```
That would disable the regular /tp-command.

The reason I was canceling the teleport earlier is because it was important for the rest of the code to know which world you just were in, in order to save that location.

I'm actually considering refactoring the `OnWorldChange` back, so I can be more lax with how teleport events are executed, as I thought I'd merge the two functions together when I refactored it into the teleport event, which obviously didn't happen.

That way, I can let teleports slip trough that are uncertain, since the position is less important than the inventory.

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/process/TeleportProcess.java:
```diff
@@ -0,0 +1,295 @@
+     * @param from   The {@link Location} they're coming from. SIDEFFECT: Might be mutated
+     * @param to     The {@link Location} they're going to. SIDEFFECT: Might be mutated
+     */
+    protected void processTeleport(Player player, PlayerTeleportEvent event, Location from, Location to) {
```
Agreed, this was the last thing I wrote yesterday, at 1AM...

SimplyLinn reviewed (commented):
I've gone trough all the major points you've made. Most of them are very sound. On a few I have a couple of questions.

MrOpposite commented:
Note to self;
TeleportProcess does not respect the teleport-settings.
Also, write tests. This proves you need them.

Commit:
Refactored back PlayerChangedWorldListener.

Commit:
Changed variable, argument and method names in Group.java. Made it use TeleportCause rather than string for its sets. Redid all documentation for teleport-highjacking to make more sense.

Commit:
Refactored rest of code to use new names of Group class. Also fixed further consistency between code and config files

Commit:
Redid the metadatamanager. Less magic, bit more verbose, easier to use.

Commit:
Minor fixes to argument names

Commit:
Minor fixes to visibility and calling static methods

Commit:
Refactored teleportprocess into multiple private methods

Commit:
Removed all String.format in debug messages

MrOpposite commented:
That should be all from the code review earlier.

If someone could give me a few pointers on how to actually write unit tests in Java/junit I'd probably do so as well. But as of now, I've never actually done anything with junit at all, and this is the first time I've touched java in years...

Commit:
Added java docstring to PWIMetaDataManager

Commit:
Added null check in teleportprocess, apparently server.getWorld(String) can return null on nonexisting worlds, but does not accept null as an argument

Gnat008 commented:
Wow, this is going to be a lot to go through. lmao. I'll try to do that later tonight or early tomorrow.

Thanks @ljacqu for your input here as well!

ljacqu commented:
I look forward to your response, @Gnat008 :)

@MrOpposite Wow, you were fast in implementing all those changes, impressive! This looks very nice now, what with the usage of TeleportCause and the more natural instantiation.
The reason I haven't replied anymore is that I want to give this PR the attention it deserves, and I still need to find a sufficiently large block of free time :smile:

MrOpposite commented:
Well, I woke up to the review, and it was a free sunday... I'll probably have less time to do stuff during the week-days, as I'm at work.

I have to say, getting back to work today was a challenge!

The amount of times I was looking for TeleportProcess.java in my project at work is embarrassing.

MrOpposite commented on src/main/resources/plugin.yml:
```diff
@@ -23,6 +23,11 @@ permissions:
+    default: true
+    children:
+      perworldinventory.useLastPosition.worldInGroup: true
+      perworldinventory.useLastPosition.locationInWorld: true
```
This is not reflecten in code.

Switch to override rather than enable for consistency with other code.

Add permissions to override within group world change location redirect

MrOpposite replied:
Ugh these names are horrible... Rename them alltogether

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/groups/Group.java:
```diff
@@ -56,9 +62,48 @@ public Group(String name, List<String> worlds, GameMode gameMode) {
+        this.enableLastPosWithinGroupCauses = enableLastPosWithinGroupCauses != null
+          ? new HashSet<>(enableLastPosWithinGroupCauses) : new HashSet<>();
+
+        this.defaultWorld = defaultWorld;
```
Make sure default world is in the world list

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/groups/Group.java:
```diff
@@ -81,6 +126,24 @@ public GameMode getGameMode() {
+     * @param defaultWorld The default world.
+     */
+    public void setDefaultWorld(String defaultWorld) {
+        this.defaultWorld = defaultWorld;
```
Make sure new default world exists in world list

MrOpposite commented on src/main/java/me/gnat008/perworldinventory/groups/Group.java:
```diff
@@ -31,6 +36,10 @@
+    private Set<TeleportCause> enableLastWorldCauses;
+    private Set<TeleportCause> enableLastPosToGroupCauses;
+    private Set<TeleportCause> enableLastPosWithinGroupCauses;
+    private String defaultWorld;
```
Make sure default world is nulled when removed from world list.

SimplyLinn reviewed (commented):
Did a read through of my PR when on the train, found a few things to fix when I get home.

Gnat008 commented:
There is a lot here to go through. Could you explain to me the whole meta data stuff? That seems way complex for just saving locations and worlds. Moreover, as far as I can tell, it isn't tied to the PWIPlayer object in any way. That object is where everything else we save about the player is located, so this should probably go there as well.

MrOpposite commented:
The Metadata is bukkits internal system for plugins to attach data to players. As long as you don't keep unnecessary references to it, it will be GC'd when a player disconnects and all that.

The reason I didn't store this in the PWIPlayer was that, at the time, I thought of the world-position data as group independent, and the PWIPlayer was tied to groups.

I realize now, on reevaluation of the idea, that the PWIPlayer actually can store all the info about the worlds in the group that PWIPlayer represents... It's actually a rather good idea and doesn't add as much complexity. I just didn't look at it from that perspective when I started working on this.

Do you want me to refactor the code to use the PWIPlayer class instead?

Gnat008 commented:
If you could, I think that would be wonderful!
